### PR TITLE
Gpu dnn pool

### PR DESCRIPTION
**Pull request by harmdevries89**
fix gh-3417

### CODE REVIEW

**Comment by nouiz** [comment]
There is conflict. This PR will need a rebase.

File: theano/sandbox/cuda/dnn.py [hunk 1]
```diff
@@ -1401,40 +1393,47 @@ class GpuDnnPool(DnnBase):
...
-        p = desc.pad
+        w = node.inputs[1]
+        s = node.inputs[2]
+        p = node.inputs[3]
+
         ret = [shape[0][0], shape[0][1],
                (shape[0][2] + 2 * p[0] - w[0]) // s[0] + 1,
                (shape[0][3] + 2 * p[1] - w[1]) // s[1] + 1]
-        if nd == 3:
+        if w.type.ndim == 3:
```
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: ndim will always be 1. this is a vector. You can use `node.inputs[0].ndim == 5`

File: theano/sandbox/cuda/dnn.py [hunk 2]
```diff
@@ -1760,13 +1788,10 @@ def dnn_pool(img, ws, stride=(1, 1), mode='max', pad=(0, 0)):
...
-    	(pad_h, pad_w) padding information.
-        pad_h is the number of zero-valued pixels added to each of the top and
-        bottom borders.
-        pad_w is the number of zero-valued pixels added to each of the left
-        and right borders.
-    nd
-    	Number of dimensions of pooling, can be 2 or 3 for 2d or 3d pooling
+        (padX, padY) padding information.
+        padX is the size of the left and right borders,
+        padY is the size of the top and bottom borders.
```
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: the old description was better.

File: theano/sandbox/cuda/tests/test_dnn.py [hunk 3]
```diff
@@ -230,6 +230,25 @@ def pool_2d_i2n(input, ds=(2, 2), strides=None, pad=(0, 0),
...
+    
+    pooled_x = cuda.dnn.dnn_pool(x, ws, stride=stride, pad=pad)
+    
+    f = theano.function([x, ws, stride, pad], pooled_x)
+    x_ = numpy.arange(500).reshape(1, 5, 10, 10).astype('float32')
+    ws_ = numpy.array([1, 1]).astype('int32')
+    stride_ = numpy.array([1, 1]).astype('int32')
+    pad_ = numpy.array([0, 0]).astype('int32')
+
+    print f(x_, ws_, stride_, pad_)
```
Comment (nouiz) [inline theano/sandbox/cuda/tests/test_dnn.py]: make the test not print something and compare again the expected value. for this, you can use the function  pool_2d_i2n() defined in that file.
Call f twice with different value of ws, stride and pad, to be sure we can change them at run time.

File: theano/sandbox/cuda/dnn.py [hunk 4]
```diff
@@ -1519,25 +1545,27 @@ def c_code(self, node, name, inputs, outputs, sub):
...
+               cudnnGetErrorString(err));
   %(fail)s
 }
-""" % dict(out=out, desc=desc, fail=sub['fail'],
+""" % dict(out=out, fail=sub['fail'],
            name=name, input=inputs[0],
-           input_desc="input"+name,
-           output_desc="output"+name)
+           ws=ws, pad=pad, str=stride,
+           nd=2, input_desc="input"+name,
```
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: nd should be `node.inputs[0].ndim - 2`

File: theano/sandbox/cuda/dnn.py [hunk 5]
```diff
@@ -1401,40 +1393,47 @@ class GpuDnnPool(DnnBase):
...
-            if img.type.ndim != e_ndim:
-                raise TypeError('img must be %dD tensor' % e_ndim)
-
-        return Apply(self, [img, desc], [img.type()])
+        
+        ws = tensor.as_tensor_variable(ws)
+        stride = tensor.as_tensor_variable(stride)
+        pad = tensor.as_tensor_variable(pad)
+        assert ws.type.ndim == stride.type.ndim and ws.type.ndim == pad.type.ndim
+        assert ws.type.ndim == 1
```
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: add an assert that img.ndim is 4 or 5.

**Comment by nouiz** [comment]
This PR need a rebase. It get too many other PR merged into it. I simply can't review it in this state.

**Comment by harmdevries89** [comment]
@nouiz I created a new PR: https://github.com/Theano/Theano/pull/3965

**Pull request closed without merging**